src/test/java/io/zentity/resolution/AbstractITCase.java

```diff
@@ -6,43 +6,65 @@
 import org.elasticsearch.client.Request;
 import org.elasticsearch.client.Response;
 import org.elasticsearch.client.RestClient;
-import org.elasticsearch.test.ESTestCase;
-import org.junit.AfterClass;
-import org.junit.BeforeClass;
+import org.testcontainers.containers.DockerComposeContainer;
+import org.testcontainers.containers.wait.strategy.Wait;
 
+import java.io.File;
 import java.io.IOException;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.time.Duration;
 
-import static org.hamcrest.Matchers.containsString;
-import static org.hamcrest.Matchers.not;
-import static org.junit.Assume.assumeThat;
-
-public abstract class AbstractITCase extends ESTestCase {
-    protected final static int HTTP_TEST_PORT = 9400;
-    protected static RestClient client;
-
-    @BeforeClass
-    public static void startRestClient() throws IOException {
-        client = RestClient.builder(new HttpHost("localhost", HTTP_TEST_PORT)).build();
-        try {
-            Response response = client.performRequest(new Request("GET", "/"));
-            JsonNode json = Json.MAPPER.readTree(response.getEntity().getContent());
-            assertTrue(json.get("tagline").textValue().equals("You Know, for Search"));
-        } catch (IOException e) {
-            // If we have an exception here, let's ignore the test
-            assumeThat("Integration tests are skipped", e.getMessage(), not(containsString("Connection refused")));
-            fail("Something wrong is happening. REST Client seemed to raise an exception.");
-            if (client != null) {
-                client.close();
-                client = null;
-            }
-        }
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.fail;
+import static org.junit.Assume.assumeFalse;
+
+public abstract class AbstractITCase {
+
+    public final static String SERVICE_NAME = "es01";
+    public final static int SERVICE_PORT = 9400;
+
+    // Create a singleton docker-compose cluster for all test classes.
+    private static final DockerComposeContainer cluster;
+    static {
+        Path path = Paths.get(System.getenv("BUILD_DIRECTORY"), "test-classes", "docker-compose.yml");
+        cluster = new DockerComposeContainer(new File(path.toString()))
+                .withEnv("BUILD_DIRECTORY", System.getenv("BUILD_DIRECTORY"))
+                .withEnv("ELASTICSEARCH_VERSION", System.getenv("ELASTICSEARCH_VERSION"))
+                .withEnv("ZENTITY_VERSION", System.getenv("ZENTITY_VERSION"))
+                .withExposedService(SERVICE_NAME, SERVICE_PORT,
+                        Wait.forHttp("/_cat/health")
+                                .forStatusCode(200)
+                                .withReadTimeout(Duration.ofSeconds(60)));
+        cluster.start();
```

Comment:
This is an awesome implementation and a really nice way to expose the service. We may want to expose a debugger as well? I've found that incredibly helpful when wanting to step through the plugin code during tests.

I implemented it by setting a [`DEBUGGER_PORT` mapping](https://github.com/fintechstudios/zentity/blob/b962fd1cf26f36126f6ae554346ae2effdef595d/docker-compose.yml#L7) and adding ` -agentlib:jdwp=transport=dt_socket,server=y,suspend=n,address=0.0.0.0:5050` to the `ES_JAVA_OPTS`.

Reply:
Can you explain that debugging process more? (e.g. after committing your example, what's the expected way to make use of it). It looks easy to add, I just haven't used debuggers in Java, and it'd be good for me to have a cursory understanding of it when supporting it.

Reply:
Yeah, sure thing! I've used the [IntelliJ `Remove JVM Debug` run configuration](https://www.jetbrains.com/help/idea/run-debug-configuration-remote-debug.html) when a debugger is exposed, which allows you to connect via the normal "debug" button and stop on breakpoints, etc.

![run-configuration-jvm-debug](https://user-images.githubusercontent.com/4655775/106531397-260dcd00-64bc-11eb-9f0c-37513401d981.png)

![debugger-running](https://user-images.githubusercontent.com/4655775/106531404-2b6b1780-64bc-11eb-9e4d-d54bdfbc189b.png)

I've found with tests, adding some arbitrary sleep before the tests start when the debugger is enabled ensures that you can connect in time.

Reply:
Sounds good, I trust your judgement so I'll implement it now and explore it later.

When debugging, do you need the docker-compose file to have every node in the cluster instrumented with the new ES_JAVA_OPTS and expose its debugger port, or does the change only apply to the one node with its Elasticsearch port exposed?

Reply:
I think what I'll do is turn this into an issue, instead of rushing it in via this PR as we get the patch out for #56. I'll want to make sure I do it right. I see more details around debugging in your fork, e.g. in AbstractITCase, that I'll need to understand. Or if you want to submit a PR for it, that works too!